profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/compute/ProfitBricksComputeServiceAdapter.java

```diff
@@ -0,0 +1,563 @@
+                    String id = String.format("cpu=%d,ram=%s,disk=%f", core, ram, size);
+                    hardwares.add(new HardwareBuilder()
+                            .ids(id)
+                            .ram(ram)
+                            .hypervisor("kvm")
+                            .name(id)
+                            .processor(new Processor(core, 1d))
+                            .volume(new VolumeImpl(size, true, true))
+                            .build());
+                }
+            }
+        }
+        return hardwares;
+    }
+
+    @Override
+    public Iterable<Provisionable> listImages() {
+        // fetch images..
+        ListenableFuture<List<Image>> images = executorService.submit(new Callable<List<Image>>() {
+
+            @Override
+            public List<Image> call() throws Exception {
+                logger.trace("<< fetching images..");
+                // Filter HDD types only, since JClouds doesn't have a concept of "CD-ROM" anyway
+                Iterable<Image> filteredImages = Iterables.filter(api.imageApi().getList(new DepthOptions().depth(1)), new Predicate<Image>() {
+
+                    @Override
+                    public boolean apply(Image image) {
+                        return image.properties().imageType() == Image.Type.HDD;
+                    }
+                });
+                logger.trace(">> images fetched.");
+
+                return ImmutableList.copyOf(filteredImages);
+            }
+
+        });
+        // and snapshots at the same time
+        ListenableFuture<List<Snapshot>> snapshots = executorService.submit(new Callable<List<Snapshot>>() {
+
+            @Override
+            public List<Snapshot> call() throws Exception {
+                logger.trace("<< fetching snapshots");
+                List<Snapshot> remoteSnapshots = api.snapshotApi().list(new DepthOptions().depth(1));
+                logger.trace(">> snapshots feched.");
+
+                return remoteSnapshots;
+            }
+
+        });
+
+        return Iterables.concat(getUnchecked(images), getUnchecked(snapshots));
+    }
+
+    @Override
+    public Provisionable getImage(String id
+    ) {
+        // try search images
+        logger.trace("<< searching for image with id=%s", id);
+        Image image = api.imageApi().getImage(id);
+        if (image != null) {
+            logger.trace(">> found image [%s].", image.properties().name());
+            return image;
+        }
+        // try search snapshots
+        logger.trace("<< not found from images. searching for snapshot with id=%s", id);
+        Snapshot snapshot = api.snapshotApi().get(id);
+        if (snapshot != null) {
+            logger.trace(">> found snapshot [%s]", snapshot.properties().name());
+            return snapshot;
+        }
+        throw new ResourceNotFoundException("No image/snapshot with id '" + id + "' was found");
+    }
+
+    @Override
+    public Iterable<DataCenter> listLocations() {
```

Comment:
Sure. Have a look at https://github.com/jclouds/jclouds/commit/7202e0557f620c272feaca3e9a49b6e45a77a791 . That commit configures the compute service to get the locations from the properties defined in the provider metadata. The important bits there are:

* The jclouds location properties configured in the provider metadata.
* The type for the locations in the generics of the adapter changes from DataCenter to the jclouds Location.
* The `LocationsFromComputeServiceAdapterModule` must be removed from the compute module. This will make the default jclouds location suppliers active and return the locations from the configured properties. You no longer need the listLocations method, so you can safely make it throw an exception.
* Since ProfitBricks needs a Datacenter, that commit declares a custom `CreateNodesInGroupThenAddToSet` strategy to take care of creating the datacenter if none exists in the selected location *before* creating any node. You can have a look at the binding and copy most of the code in that strategy (if not all).
* To let the datacenter be created *before* the adapter is called, that commit introduces an internal class that adds the created DataCenter to the jclouds Template, so it can be directly passed to the adapter, removing the need to perform additional calls.

Feel free to ask if you have any doubts or need some more guidance!